x/or/da/keeper/cc.go

```diff
@@ -1,6 +1,326 @@
+		}
+		if frame.Header.GetL1Height() < ccState.GetL1Height() {
+			return types.ErrInvalidCCBatch.Wrapf("frame %d is outdated: %d (frame) < %d (CCState)", i, frame.Header.GetL1Height(), ccState.GetL1Height())
+		}
+
+		for j, elem := range frame.Elements {
+			if elem.Txraw == nil {
+				if elem.QueueIndex < 1 || elem.QueueIndex != queueState.ProcessedQueueIndex+1 {
+					return types.ErrInvalidCCBatch.Wrapf("queue index of frame %d element %d is %d, expected %d", i, j, elem.QueueIndex, queueState.ProcessedQueueIndex+1)
+				}
+
+				qtx, err := k.GetQueueTx(ctx, rollupName, elem.QueueIndex)
+				if err != nil {
+					return err
+				}
+				qtx.Status = types.QUEUE_TX_SUBMITTED
+				k.saveQueueTx(ctx, rollupName, elem.QueueIndex, qtx)
+
+				queueState.ProcessedQueueIndex++
+			}
+		}
+
+		ccState.ProcessedL2Block++
+		ccState.Timestamp = frame.Header.Timestamp
+		ccState.L1Height = frame.Header.L1Height
+		k.setL2HeightBatchMap(ctx, rollupName, ccState.ProcessedL2Block, ccState.Height)
+	}
+
+	k.setCCRef(ctx, rollupName, ccState.Height, ref)
+	k.setCCState(ctx, rollupName, ccState)
+	k.setQueueTxState(ctx, rollupName, queueState)
+
+	if err := ctx.EventManager().EmitTypedEvent(&types.EventAppendCCBatch{
+		RollupName:          rollupName,
+		BatchIndex:          ccState.Height,
+		ProcessedQueueIndex: queueState.ProcessedQueueIndex,
+		TotalFrames:         ref.TotalFrames,
+		BatchSize:           ref.BatchSize,
+		BatchHash:           ref.BatchRoot,
+		ProcessedL2Block:    ccState.ProcessedL2Block,
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k Keeper) UpdateQueueTxsStatus(ctx sdktypes.Context) error {
+	rollupList := k.rollupKeeper.GetRegisteredRollups(ctx)
+	if rollupList == nil {
+		return nil
+	}
+
+	for _, name := range rollupList {
+		state, err := k.GetQueueTxState(ctx, name)
+		if err != nil {
+			continue
+		}
+
+		if err := k.processQueueTxs(ctx, state, name); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (k Keeper) processQueueTxs(ctx sdktypes.Context, state *types.QueueTxState, name string) error {
+	qi := state.ProcessedQueueIndex + 1
+	for ; qi < state.NextQueueIndex; qi++ {
+		qtx, err := k.GetQueueTx(ctx, name, qi)
+		if err != nil {
+			return types.ErrQueueTxStateNotFound.Wrapf("rollup %s queue tx index %d", name, qi)
+		}
+
+		if (uint64(qtx.L1Height) + k.QueueTxExpirationWindow(ctx)) > uint64(ctx.BlockHeight()) {
+			break
+		}
+
+		if qtx.Status == types.QUEUE_TX_PENDING {
+			qtx.Status = types.QUEUE_TX_EXPIRED
+			k.saveQueueTx(ctx, name, qi, qtx)
+			// TODO: slash registered sequencers
+		}
+	}
+	state.ProcessedQueueIndex = qi - 1
+	k.setQueueTxState(ctx, name, state)
+
+	return nil
+}
+
+func (k Keeper) DecompressCCBatch(ctx sdktypes.Context, origin types.CompressedCCBatch) (*types.CCBatch, error) {
+	p := k.GetParams(ctx)
+	if uint64(len(origin.Data)) > p.CCBatchMaxBytes {
```

Comment:
Is CCBatchMaxBytes a limit for compressed data?

Reply:
Yes, because that's actually the data stored in IAVL.